Add unittest framework

stin7 opened the pull request:
I created a new branch for unit test suite, per issue #21. New code resides in folder outside of `backend`, so this branch should be safe to bring into master.

I decided to submit this pull request before all tests are written (most are not) so that others can help build them out.

stin7 committed:
add test runner script and unit tests folder

stin7 committed:
started test_User.py

stin7 committed:
moved unit tests out of backend

stin7 committed:
tests for User class

stin7 committed:
fixed indentation

stin7 committed:
moved testrunner to root so it can access /backend for import

stin7 committed:
pointed testrunner to unit test folder

stin7 committed:
added __init__.py to backend so test runner could import folder as a module

stin7 committed:
added blurb on Unittests to README

stin7 committed:
changed test file to be more generic (file focused not class focused)

stin7 committed:
added placeholders for tests

stin7 committed:
added option to override SDK_PATH

stin7 committed:
added stripe to path so module can be imported

stin7 committed:
added placeholders for other tests in main.py

stin7 committed:
fixed test for User.createOrUpdate

stin7 commented:
Quick follow up - I misspoke, there was one change to `backend`: I added a blank `__init__.py` file.

Also, I thought my pull request would be for a new branch called `add-unittest-framework` but it looks like this is set to go straight into `master`. Let me know if I should redo anything on my end, I'm fairly new to Forking and Pull Requests on Github.

Cheers,
Austin

brad commented:
It looks like you did it right, although it's unfortunately not merge-able right now. I'm guessing it's a conflict in README.md. If you get a chance, try and merge MayOneUS/master into your branch and see if you can resolve the conflicts. I'll take a look at this as soon as I get a chance.

stin7 committed:
Merge branch 'master' of https://github.com/MayOneUS/app-engine into add-unittest-framework

Conflicts:
	README.md

stin7 commented:
It was README. I pulled the current master into my branch and resolved the conflict (dropped my changes)

If this pull request uses the current state of the branch it should be good to merge. If not I can resubmit a new pull request.

Cheers,
Austin

stin7 committed:
minor cleanup

stin7 committed:
added test_Model.py to reflect changes in /backend

stin7 committed:
updated test_Main.py to reflect changes in /backends/main.py

brad commented on testrunner.py:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/python
+import sys
+# Install the Python unittest2 package before you run this script.
```
The `unittest` module is built into python, isn't it? I believe this comment is misleading.

brad commented on testrunner.py:
```diff
@@ -0,0 +1,30 @@
+  unittest.TextTestRunner(verbosity=2).run(suite)
+
+if __name__ == '__main__':
+  print('UNITTESTS FOR MAYONE APPENGINE')
```
I don't think this or the output on line 29 are necessary, the tests can speak for themselves.

brad commented on testrunner.py:
```diff
@@ -0,0 +1,30 @@
+  print('UNITTESTS FOR MAYONE APPENGINE')
+  SDK_PATH = SDK_PATH_manual
+  TEST_PATH = TEST_PATH_manual
+  SDK_PATH_override = raw_input('Enter SDK_PATH Override (Leave blank to use the default): ')
```
I would much prefer if this variable was obtained from `sys.argv`. That way the tests can be ran unattended.

brad commented on unittests/test_Main.py:
```diff
@@ -0,0 +1,86 @@
+from backend import main
+
+class TestGetTotalHandler(unittest.TestCase):
+    def setUp(self):
```
Please stick with a two space indent consistently throughout this file.

brad commented on unittests/test_Main.py:
```diff
@@ -0,0 +1,86 @@
+      self.testbed.deactivate()
+     
+    def test_get(self):
+      self.fail('TEST NOT IMPLEMENTED')
```
I don't feel great about committing a bunch of failing tests. Can we leave them out until they do something/pass?

erinzm replied:
No, do commit failing tests. They are a big part of testing; as a template, they help you write code to _make the tests pass_.

brad replied:
OK, sure, I'm with you. It's good to have that encouragement to get them done.

hjfreyer replied:
FWIW, I disagree. I do TDD, but you don't break the branch for others.

On Wednesday, May 14, 2014 1:18:32 PM, Brad Pitcher <
notifications@github.com> wrote:

> In unittests/test_Main.py:
> 
> > +from backend import main
> > +
> > +class TestGetTotalHandler(unittest.TestCase):
> > -    def setUp(self):
> > -      # First, create an instance of the Testbed class.
> > -      self.testbed = testbed.Testbed()
> > -      # Then activate the testbed, which prepares the service stubs for use.
> > -      self.testbed.activate()
> > -      # Next, declare which service stubs you want to use.
> > -      self.testbed.init_datastore_v3_stub()
> >   +
> > -    def tearDown(self):
> > -      self.testbed.deactivate()
> >   +
> > -    def test_get(self):
> > -      self.fail('TEST NOT IMPLEMENTED')
> 
> OK, sure, I'm with you. It's good to have that encouragement to get them
> done.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/MayOneUS/app-engine/pull/41/files#r12648455
> .

brad commented on unittests/test_Model.py:
```diff
@@ -0,0 +1,154 @@
+
+
+class TestPledge(unittest.TestCase):
+    def setUp(self):
```
Again, please stick with a two space indent throughout this file.

brad commented:
@stin7 I have made some inline comments requesting changes. But this looks really great. I ran the tests locally, everything ran smoothly.

stin7 committed:
removed misleading comment

stin7 committed:
removed unneeded print out

stin7 committed:
covered 4 space indent to 2 space indent

stin7 committed:
pulling override from first argument in sys.argv instead of a command prompt

stin7 committed:
removed extraneous print out

stin7 commented:
Great - I've fixed the indentation, removed extraneous comments and print outs, and am pulling the first passed argument (if it exists) as an override for the SDK_PATH

brad commented:
You know, Hunter is right. I don't know if you saw his inline comment, but let's just start with the working tests. Failing tests are good on your local machine, but they shouldn't affect the rest of us. For the why, imagine the following scenario: I am working on feature X, so I write tests to go along with it. If there are still failing tests after I get my tests working, I might think my feature had an inadvertent effect on the code base and go searching for a way to fix it.
No, better to just keep the failing tests to yourself and commit them when they pass.

stin7 commented:
Sounds good to me - will remove the fail assertions. Should they be replaced with a logged warning to remind us of incomplete coverage? Or just leave it to us to remember and get to full coverage by some other means? I'm fine with either.

brad commented:
Just remove them completely for now. @hjfreyer @jtolds  I've got some experience with Travis ci and coveralls.io. What do you think about integrating with those services to have some continuous integration with coverage reports?

erinzm commented:
Sounds good; does anyone want Travis/Coveralls integration with IRC #mayone?

jtolio commented:
that sounds awesome @brad and @ArchimedesPi. love me those build|passing tags :)

erinzm commented:
Let's do Travis; write a make/rakefile that runs the tests, pushes to Coveralls, grabs the local coveralls results, and connects to `#mayone` and posts up the tests as `@travis`. Sound good?
(edit: there's a Coveralls/Travis framework for python tests here: https://github.com/coagulant/coveralls-python)

brad commented:
Yeah except it can be done in .travis.yml so there's no need for a
[mr]akefile. I'm thinking it is a separate issue from this one. Let's get
this taken care of first.
On May 15, 2014 2:19 PM, "Liam M" notifications@github.com wrote:

> Let's do Travis; write a make/rakefile that runs the tests, pushes to
> Coveralls, grabs the local coveralls results, and connects to #mayone and
> posts up the tests as @travis. Sound good?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/MayOneUS/app-engine/pull/41#issuecomment-43245927
> .

stin7 committed:
commented out non-implemented tests.

stin7 committed:
Revert "commented out non-implemented tests."

This reverts commit 37e9abb3c32fa68fe483b7e7a01d19f4d361a417.

stin7 committed:
removed non implemented tests

stin7 committed:
fixed indentation

stin7 commented:
LMK if there's anything else I can do on this

brad commented:
Thank you @stin7 !
Looks great
